commons/src/main/java/esa/commons/io/AbstractPathWatcher.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2022 OPPO ESA Stack Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package esa.commons.io;
+
+import esa.commons.Checks;
+import esa.commons.concurrent.ConcurrentHashSet;
+import esa.commons.logging.Logger;
+import esa.commons.logging.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+abstract class AbstractPathWatcher implements PathWatcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPathWatcher.class);
+    private final Path path;
+    private final ArrayList<WatchEvent.Kind<?>> events = new ArrayList<>(4);
+    private final Consumer<WatchEventContext<?>> create;
+    private final Consumer<WatchEventContext<?>> delete;
+    private final Consumer<WatchEventContext<?>> modify;
+    private final Consumer<WatchEventContext<?>> overflow;
```

Comment:
specific event types should correspond to specific contexts instead of contexts with `?`